### PR TITLE
Adds the V1 Version of the DSL

### PR DESCRIPTION
**Pull request by VincenzoFerme**

### CODE REVIEW

**Comment by VincenzoFerme** [comment]
@igormoreno Refer to Issue #59 to start creating the BenchFlow Codacy configuration file, starting from the projects we currently have that are mainly Scala and Java. You can update this pull request with the configuration file, so it starts applying to the code committed here, and it is going to be applied to all the current and new projects that have to be updated accordingly.

File: benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/simone/BenchFlowExperiment.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                               sut: Sut,
+                               users: Users,
+                               drivers: Seq[Driver[_ <: Operation]],
+                               trials: TotalTrials,
+                               execution: LoadFunction,
+                               properties: Properties,
+                               sutConfiguration: SutConfiguration)
+{
+  def getAliasForService(serviceName: String) = sutConfiguration.deployment.get(serviceName)
+  // TODO - move me
```
Comment (VincenzoFerme) [inline benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/simone/BenchFlowExperiment.scala:25]: Without these methods, the DSL is not useful in BenchFlow

File: benchflow-dsl/src/test/scala/cloud/benchflow/dsl/definition/types/time/TimeTest.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import net.jcazevedo.moultingyaml._
+import cloud.benchflow.dsl.definition.types.time.TimeYamlProtocol._
+
+import scala.util.Try
+
+/**
+  * @author Jesper Findahl (jesper.findahl@usi.ch) 
+  *         created on 14.03.17.
+  */
+class TimeTest extends AssertionsForJUnit {
```
Comment (VincenzoFerme) [inline benchflow-dsl/src/test/scala/cloud/benchflow/dsl/definition/types/time/TimeTest.scala]: Missing `micro` and check that you include all

**Comment by VincenzoFerme** [comment]
Remaining issues:

- semantic validation https://github.com/benchflow/benchflow/issues/48
- cleanup Simone's code, and what to be taken from it https://github.com/benchflow/benchflow/issues/103
- not defined elements (e.g., strategy) https://github.com/benchflow/benchflow/issues/104
- define defaults value, when feasible and makes sense https://github.com/benchflow/benchflow/issues/105
- which types of tests we decided to implement, and what is missing here https://github.com/benchflow/benchflow/issues/106

and reply to a comment to this pull request to link them as missing here. 


Moreover:

- How can I get an experiment representation? https://github.com/benchflow/benchflow/issues/107
- Where are the different mixes definitions that were in https://github.com/benchflow/benchflow/pull/67/files#diff-759278330151bb082f8a28c58c208570

File: benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/simone/BenchFlowExperiment.scala (hunk 1 above)
Comment (jeppe-style) [reply benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/simone/BenchFlowExperiment.scala:25]: If we need helper methods I think that is fine, however, we can get it also directly from the BenchFlowTest/Experiment instance.

**Comment by jeppe-style** [comment]
Mixes are defined here: https://github.com/benchflow/benchflow/tree/feature-dsl-v1-initial/benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/workload/mix

issue regarding experiment representation: https://github.com/benchflow/benchflow/issues/107

other issues:
- semantic validation https://github.com/benchflow/benchflow/issues/48
- cleanup Simone's code, and what to be taken from it https://github.com/benchflow/benchflow/issues/103
- not defined elements (e.g., strategy) https://github.com/benchflow/benchflow/issues/104
- define defaults value, when feasible and makes sense https://github.com/benchflow/benchflow/issues/105
- which types of tests we decided to implement, and what is missing here https://github.com/benchflow/benchflow/issues/106

File: benchflow-dsl/src/test/scala/cloud/benchflow/dsl/definition/types/time/TimeTest.scala (hunk 2 above)
Comment (jeppe-style) [reply benchflow-dsl/src/test/scala/cloud/benchflow/dsl/definition/types/time/TimeTest.scala]: Updated in commit https://github.com/benchflow/benchflow/pull/67/commits/d1b3f4c4ca1357fbcc2031a3b776a86ca1ddf240

File: benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/simone/BenchFlowExperiment.scala (hunk 1 above)
Comment (VincenzoFerme) [reply benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/simone/BenchFlowExperiment.scala:25]: It is fine to have them in the experiment

**Review by igormoreno — APPROVED** [review]
I understand that for a 1st version of the DSL a lot off "set up"/"architecture" code is necessary but I would advise that future pull requests should be much smaller and focused around a given subject.

I already talked to @jeppe-style about how to improve the tests and a good way to replace the current uses of "var" and the current calls to "get" on optional types (Try and Option). I think these changes should be incorporated in a future pull request which should also get rid of the current uses of "???".

**Comment by VincenzoFerme** [comment]
@igormoreno I agree with you, i told the same to @jeppe-style because of the difficulties to review such big pull requests. For this pull request we agreed it is ok.

@jeppe-style can you link to the issues discussing what @igormoreno pointed out in https://github.com/benchflow/benchflow/pull/67#pullrequestreview-28423957

**Comment by VincenzoFerme** [comment]
@igormoreno I had a look at the `scalastyle` configuration file you added. We also need https://support.codacy.com/hc/en-us/articles/115002130625-Codacy-Configuration-File/, we need to document the settings. After we verify it works, we can also add other Scala checkers configuration as well as configuring the Java tools.

File: benchflow-dsl/pom.xml [hunk 3]
```diff
@@ -175,6 +175,30 @@
...
+              <version>0.8.0</version>
+              <configuration>
+                <verbose>false</verbose>
+                <failOnViolation>true</failOnViolation>
+                <includeTestSourceDirectory>true</includeTestSourceDirectory>
+                <failOnWarning>false</failOnWarning>
+                <sourceDirectory>${basedir}/src/main/scala</sourceDirectory>
+                <testSourceDirectory>${basedir}/src/test/scala</testSourceDirectory>
+                <configLocation>${basedir}/../scalastyle-config.xml</configLocation>
+                <outputFile>${project.basedir}/scalastyle-output.xml</outputFile>
```
Comment (VincenzoFerme) [inline benchflow-dsl/pom.xml:191]: @igormoreno you might need to add this file in `.gitignore`

**Comment by jeppe-style** [comment]
@VincenzoFerme @igormoreno in reference to https://github.com/benchflow/benchflow/pull/67#pullrequestreview-28423957
- the issue (https://github.com/benchflow/benchflow/issues/58) describing what we should change to replace `var` with `val` (example is in the code)
- the issue (https://github.com/benchflow/benchflow/issues/106) describing what tests are missing

**Comment by jeppe-style** [comment]
I worked on the code review. Most of the things are regarding indentation or things we had already talked about changing, or code in the package `simone`. Most of the severe issues are incorporated in a new commit (below).

- [YamlErrorHandler - Call-by-name parameter should be in last position](https://www.codacy.com/app/BenchFlow/benchflow/file/5831089621/issues/source?fileBranchId=4326789#l@result.result.startLine). It says to change the order of the parameters. It would mean changes in ~50 places. Ok to ignore?
- [Time - Covariant equals without overriding equals(java.lang.Object).](https://www.codacy.com/app/BenchFlow/benchflow/file/5831089654/issues/source?bid=4283320&fileBranchId=4326789#l18). This seems to be not correct since the parameter is `scala.Any`

**Comment by jeppe-style** [comment]
We are merging! 
@VincenzoFerme please carefully check that all the remaining work is mentioned in issues, thanks

**Pull request merged**